### PR TITLE
support az aware hashring and multiple sts in one hashring

### PR DESCRIPTION
**Pull request by christopherzli**
## What changes are proposed in this pull request?
1. Support az aware hashring in Thanos v0.32+ 
https://thanos.io/tip/components/receive.md/#az-aware-ketama-hashring-experimental
2. supported multiple sts in one hashring
3. Update dependencies including golang version, along with bingo and golangci-lint version, and fixed all lint tests
## How is this tested?
main_test.go,
Make thanos-receive-controller
https://github.com/observatorium/thanos-receive-controller/issues/127

### CODE REVIEW

File: .bingo/golangci-lint.mod [hunk 1]
```diff
@@ -2,4 +2,4 @@ module _ // Auto generated by https://github.com/bwplotka/bingo. DO NOT EDIT
 
 go 1.17
 
-require github.com/golangci/golangci-lint v1.45.2 // cmd/golangci-lint
```
Comment (christopherzli) [inline .bingo/golangci-lint.mod:5]: needs to update golangci-lint version and bingo version to work with golang v1.21

File: .golangci.yml [hunk 2]
```diff
@@ -19,6 +19,11 @@ linters:
   - varnamelen
   - paralleltest
   - cyclop
+  - exhaustruct
```
Comment (christopherzli) [inline .golangci.yml:22]: temp disable some newly added lint checks, feel free to resolve on these lint issues

File: go.mod [hunk 3]
```diff
@@ -1,116 +1,164 @@
...
+	github.com/go-kit/log v0.2.1
+	github.com/google/go-cmp v0.6.0
 	github.com/oklog/run v1.1.0
-	github.com/prometheus/client_golang v1.12.1
-	github.com/thanos-io/thanos v0.25.2
-	k8s.io/api v0.26.1
-	k8s.io/apimachinery v0.26.1
-	k8s.io/client-go v0.26.1
+	github.com/prometheus/client_golang v1.17.0
+	github.com/thanos-io/thanos v0.33.0
```
Comment (christopherzli) [inline go.mod:12]: accomodate thanos newer version

File: main.go [hunk 4]
```diff
@@ -292,35 +298,35 @@ func splitLabel(in string) (key, value string) {
 	return parts[0], parts[1]
 }
 
-func (p prometheusReflectorMetrics) NewListsMetric(name string) cache.CounterMetric {
+func (p prometheusReflectorMetrics) NewListsMetric(_ string) cache.CounterMetric {
 	return p.listsMetric
 }
 
-func (p prometheusReflectorMetrics) NewListDurationMetric(name string) cache.SummaryMetric {
+func (p prometheusReflectorMetrics) NewListDurationMetric(_ string) cache.SummaryMetric {
```
Comment (christopherzli) [inline main.go:305]: fix lint issues

File: main.go [hunk 5]
```diff
@@ -283,7 +289,7 @@ func newReflectorMetrics(reg *prometheus.Registry) prometheusReflectorMetrics {
 
 const labelParts = 2
 
-func splitLabel(in string) (key, value string) {
+func splitLabel(in string) (string, string) {
```
Comment (christopherzli) [inline main.go:292]: fix nonamed return values

File: main.go [hunk 6]
```diff
@@ -579,7 +587,13 @@ func (c *controller) sync(ctx context.Context) {
 		}
 
 		c.replicas[hashring] = *sts.Spec.Replicas
-		statefulsets[hashring] = sts.DeepCopy()
+
+		if _, ok := statefulsets[hashring]; !ok {
+			statefulsets[hashring] = []*appsv1.StatefulSet{}
+		}
+		// Append the new value to the slice associated with the hashring key
+		statefulsets[hashring] = append(statefulsets[hashring], sts.DeepCopy())
```
Comment (christopherzli) [inline main.go:595]: support multiple sts placed in a single hashring in our setup

File: main.go [hunk 7]
```diff
@@ -679,6 +686,40 @@ func (c *controller) populate(ctx context.Context, hashrings []receive.HashringC
 	}
 }
 
+func (c *controller) populateEndpoint(sts *appsv1.StatefulSet, podIndex int, err error, pod *corev1.Pod) *receive.Endpoint {
```
Comment (christopherzli) [inline main.go:689]: extracted this function to reduce cognitive complexity required by lint

**Comment by christopherzli** [comment]
tested it worked when scale up 1 node, the configmap updates the configmap too: 
Screenshot 2024-01-25 at 2 16 19 AM

**Review by philipgough — APPROVED** [review]
lgtm thanks for this

**Pull request merged**